### ISSUE TITLE
[Feature Question] Frame caching for smooth stepping / jog shuttle

### ISSUE THREAD

**Issue by cajhin** [clsid2/mpc-hc#2286]
Thank you so much for maintaining the best player, especially updating outdated things like subtitle sources. it would be awful to lose MPC.

Now I get it, I have to implement new features myself (or sponsor someone).
Before I dig in, I would like to ask if I have a chance to succeed.

**The Goal**
I often make screenshots of "the perfect frame". I would love fluid frame stepping (or jog shuttle), the way video editors work.
Among video *players*, MPC is the best I know, been using it since the Gabest days.
This https://zurbrigg.com/keyframe-mp is framestep heaven, but it's no good as a general player.

The biggest issue is frame stepping backwards; I believe that MPC has to decode all the delta frames from the last keyframe to the current frame, every time you step back, again and again.
The obvious solution for that would be to cache all relevant frames.

**The Challenge**
I can code, some C++ too if I have to, I'm not worried much about implementing an image cache, BUT
I have zero knowledge about how the video processing pipeline works, it's all a super complicated black box to me, and I don't want to spend major effort understanding everything in detail.

**The Question**
Is there an easy way to hook into the video pipeline, grab the decoded frames, and push frames back into the pipeline?
Basically, on framestep event, I would first check if the new frame is in cache, and only on cache miss continue with the regular 'decode video, get new frame'.
The important part is that I don't want to deal with video, I want to juggle with nice, easy images.
I looked at the code, but didn't find anything to hook into.

**Comment by clsid2** [comment]
DirectShow is not suitable for framestep backwards.

I have zero interest in this subject.
